added simple arty calculator

NeutraleNull opened the pull request:
Implemented my idea from #2 
![20220928020216_1](https://user-images.githubusercontent.com/18582939/192658763-78f8fe41-141c-42d8-b17d-95f1eb39503c.jpg)

It works together with the artillery table, or more say more clearly, it does not work without one.
First you select your charge from the table, then head over to self interact -> easy arty -> start. 
Just click on the map and a solution will be shown if available. Otherwise a error message will be shown. 
You can switch the charge and your own position and the solution will be updated accordingly.

NeutraleNull committed:
added simple arty calculator

NeutraleNull commented:
Code can still need some improvements here and there (the estimation of the elevation between two settings is not that great).

diwako commented on scripts/easyarty/funcs.hpp:
```diff
@@ -0,0 +1,12 @@
+    tag = "easyarty";
+    class functions {
+        file = "scripts\easyarty\functions";
+        class preInit{preInit = 1;};
```
If preinit is not needed, do not include

diwako commented on scripts/easyarty/functions/fn_calculate.sqf:
```diff
@@ -0,0 +1,46 @@
+
+private _az = (_vehicle getDir _targetPos) * DEGREE_TO_MIL_FACTOR;
+
+_distanceToTarget = _vehicle distance artyTarget;
```
missing private keywords for new variable declarations

diwako commented on scripts/easyarty/functions/fn_postInit.sqf:
```diff
@@ -0,0 +1,64 @@
+if (isDedicated) exitWith {};
```
use this instead, covers headless client as well then
```sqf
if !(hasinterface) exitWith {};
```

diwako commented on scripts/easyarty/functions/fn_postInit.sqf:
```diff
@@ -0,0 +1,64 @@
+if (isDedicated) exitWith {};
+easyArtyRunning = false;
```
heavily suggesting the use of proper prefixing
easyarty_running
easyarty_pfh
easyarty_displayEh

this means it becomes less problematic moving the script to other missions or even a mod some time in the future

diwako commented on scripts/easyarty/functions/fn_postInit.sqf:
```diff
@@ -0,0 +1,64 @@
+easyArtyDisplayEh = -1;
+
+private _spacerAction = [
+    "spacerEasyArty",
```
use prefixes for the action IDs

diwako commented on scripts/easyarty/functions/fn_postInit.sqf:
```diff
@@ -0,0 +1,64 @@
+    1,
+    ["ACE_SelfActions"],
+    _spacerAction
+] call ace_interact_menu_fnc_addActionToObject;
```
use `ace_interact_menu_fnc_addActionToClass` instead, else a respawn will break this.

diwako commented on scripts/easyarty/functions/fn_setTargetPos.sqf:
```diff
@@ -0,0 +1,17 @@
+	params ["", "_pos"];
+	deleteMarker "target_loc";
+	_marker = createMarkerLocal ["target_loc", _pos];
+	_marker setMarkerType "HD_DOT";
```
this will make the marker global, i doubt you want this. Also use prefixing for the marker name.

diwako commented on scripts/easyarty/functions/fn_setTargetPos.sqf:
```diff
@@ -0,0 +1,17 @@
+private _mapClickEh = addMissionEventHandler ["MapSingleClick", {
+	params ["", "_pos"];
+	deleteMarker "target_loc";
+	_marker = createMarkerLocal ["target_loc", _pos];
```
missing private keyword

diwako commented on scripts/easyarty/functions/fn_setTargetPos.sqf:
```diff
@@ -0,0 +1,17 @@
+
+private _mapClickEh = addMissionEventHandler ["MapSingleClick", {
+	params ["", "_pos"];
+	deleteMarker "target_loc";
```
global delete of a none global marker.

NeutraleNull committed:
updated code as suggested in PR review

NeutraleNull commented:
Updated the code according to @diwako suggestions. Many thanks for the review :)
Also added back in the functionality to switch between low and high shell arc.

NeutraleNull committed:
amend missing change

NeutraleNull committed:
fixed shells landing short and included tot to view

NeutraleNull commented:
Fixed shells landing short and included time on target to view.
Here a little video showing how it works, and how precise it can be:
https://youtu.be/mqoG4zK83Xw

b-mayr-1984 commented:
> Here a little video showing how it works, and how precise it can be: https://youtu.be/mqoG4zK83Xw

I tried it and immediately got confused. After I went in game I immediately tried the new self ACE menu point.
![image](https://user-images.githubusercontent.com/76476468/194711863-10f1306b-d529-457f-8b35-5bee84e94385.png)
This message showed up which did not explain to me what the problem was or how I should even come up with a solution.
![image](https://user-images.githubusercontent.com/76476468/194711873-8c7b046a-72ee-48fa-b775-7dcabe63b377.png)
Only in connection with your YT video I managed to make progress by clicking in the range card (without knowing what I am doing).

I also noticed that the Easy Arty ACE menu items are visible if you have no range cards with you. To my understanding Easy Arty can not work without range cards and if so should not be made available in the first place.

NeutraleNull commented on scripts/easyarty/functions/fn_postInit.sqf:
```diff
@@ -0,0 +1,69 @@
+    "Easy Arty",
+    "",
+    {},
+    {true}
```
tobe added [player, "ACE_artilleryTable"] call BIS_fnc_hasItem instead of true

NeutraleNull committed:
check if player has an artillery table

check if a player has an artillery table before adding the spacer action to his player

NeutraleNull commented:
@b-mayr-1984 wie ist der Stand der Dinge hier?

b-mayr-1984 commented:
> @b-mayr-1984 wie ist der Stand der Dinge hier?

Ich hab hier auf den Input gewartet den du mir noch geben wolltest.
https://discord.com/channels/@me/1000698946135543809/1028337333839351920

Hab dann einige Wochen später nochmal nachgefragt bei dir, aber leider keine Antwort erhalten. 
https://discord.com/channels/@me/969958150377730119/1055892275424665600

Daher ist aus meiner Sicht gerade der Stand **on hold**. 

Um das zu mergen bräuchte es mMn konsistentes Verhalten, denn sonst sehe ich Spielerfrust vorprogrammiert wenn es nur in manchen Fällen funktioniert.

Daher nochmal die Frage:
@NeutraleNull willst du hier noch was tun um das zu stabilisieren bzw. konsistent zu machen?
Wenn die Behebung sich zu schwierig gestalten würde, dann könnte man es für die Spieler im User Interface evtl. als dediziertes **Experimental Feature** deklarieren, welches **known issues** hat (macht RHS towing ja auch).

NeutraleNull commented:
Alles klar, ich würde es Stable für Arty nennen. Mörser funktionieren ja damit -noch- nicht. Schaue ich mir aber auf jeden Fall mal an.

b-mayr-1984 commented:
> Alles klar, ich würde es Stable für Arty nennen. Mörser funktionieren ja damit -noch- nicht. Schaue ich mir aber auf jeden Fall mal an.

Passt. Solange die Nutzerführung das mit fehlendem Mörser-Support widerspiegelt (z.B. durch entsprechendes benennen der ACE-Menüeinträge), soll mir (und hoffentlich den Spielern) das reichen.

Wenn du Mörser auch noch an's Laufen kriegen möchtest, sogar noch besser. 🙂
